csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Assets.cs

```diff
@@ -0,0 +1,232 @@
+
+namespace Semmle.Extraction.CSharp.DependencyFetching
+{
+    /// <summary>
+    /// Class for parsing project.assets.json files.
+    /// </summary>
+    internal class Assets
+    {
+        private readonly ProgressMonitor progressMonitor;
+
+        private readonly string[] netFrameworks = new[] {
+            "microsoft.netcore.app.ref",
+            "microsoft.netframework.referenceassemblies"
+        };
+
+
+        internal Assets(ProgressMonitor progressMonitor)
+        {
+            this.progressMonitor = progressMonitor;
+        }
+
+        /// <summary>
+        /// In some cases the paths in asset files are using a wildcard like pattern.
+        /// If the path doesn't point to a specific DLL, we include the entire directory instead.
+        /// </summary>
+        private static string ParseFilePath(string path)
+        {
+            if (!path.EndsWith(".dll"))
+            {
+                return Path.GetDirectoryName(path) ?? path;
+            }
+            return path;
+        }
+
+        /// <summary>
+        /// Class needed for deserializing parts of an assets file.
+        /// It holds information about a reference.
+        /// </summary>
+        private class ReferenceInfo
+        {
+            // This carries the type of the reference.
+            // We are only interested in package references.
+            public string Type { get; set; } = "";
+
+            // If not a .NET framework reference we assume that only the files mentioned
+            // in the compile section are needed for compilation.
+            public Dictionary<string, object> Compile { get; set; } = new Dictionary<string, object>();
+        }
+
+        /// <summary>
+        /// Gets the package dependencies from the assets file.
+        /// 
+        /// Parse a part of the JSon assets file and returns the paths
+        /// to the dependencies required for compilation.
+        ///
+        /// Example:
+        /// {
+        ///   "Castle.Core/4.4.1": {
+        ///     "type": "package",
+        ///     "compile": {
+        ///       "lib/netstandard1.5/Castle.Core.dll": {
+        ///         "related": ".xml"
+        ///        }
+        ///     }
+        ///   },
+        ///   "Json.Net/1.0.33": {
+        ///     "type": "package",
+        ///     "compile": {
+        ///       "lib/netstandard2.0/Json.Net.dll": {}
+        ///     },
+        ///     "runtime": {
+        ///       "lib/netstandard2.0/Json.Net.dll": {}
+        ///     }
+        ///   }
+        /// }
+        /// 
+        /// Returns {
+        ///   "castle.core/4.4.1/lib/netstandard1.5/Castle.Core.dll",
+        ///   "json.net/1.0.33/lib/netstandard2.0/Json.Net.dll"
+        /// }
+        /// </summary>
+
+        private IEnumerable<string> GetPackageDependencies(JObject json)
+        {
+            // If there are more than one framework we need to pick just one.
+            // For now, we just pick the first one.
```

Comment:
I think it's okay to pick one, but we should make sure that the order is stable, and we always choose the same item.

Reply:
That is a good point. There is not much documentation of project.assets.json generation - so I can't say for sure that it is stable. However, we can assume that the listed frameworks are stable - I will just take the "max" using lexicographic ordering.